### PR TITLE
React to Routing repo addition of constraints

### PR DESCRIPTION
**Pull request by kichalla**
Related to https://github.com/aspnet/Routing/pull/584

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs [hunk 1]
```diff
@@ -239,7 +248,7 @@ private bool MatchRouteValue(ActionDescriptor action, MvcEndpointInfo endpointIn
                     //
                     // REVIEW: This is really ugly
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs]: Can remove ugliness comment?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs [hunk 2]
```diff
@@ -239,7 +248,7 @@ private bool MatchRouteValue(ActionDescriptor action, MvcEndpointInfo endpointIn
...
                     if (endpointInfo.Constraints.TryGetValue(routeKey, out var constraint)
-                        && !constraint.Match(new DefaultHttpContext() { RequestServices = _serviceProvider }, new DummyRouter(), routeKey, new RouteValueDictionary(action.RouteValues), RouteDirection.IncomingRequest))
+                        && !constraint.Match(routeKey, new RouteValueDictionary(action.RouteValues), RouteDirection.IncomingRequest))
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs]: Minor optimization but is there a way to reuse a RVD of `action.RouteValues`?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs [hunk 3]
```diff
@@ -358,6 +369,38 @@ private void EnsureRequiredValuesInDefaults(RouteValueDictionary requiredValues,
             }
         }
 
+        private IDictionary<string, IEndpointMatchConstraint> GetConstraints(
+            RouteTemplate parsedTemplate,
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs]: Debug.Assert that parsedTemplate is not null

File: src/Microsoft.AspNetCore.Mvc.Core/Routing/KnownRouteValueEndpointMatchConstraint.cs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                throw new ArgumentNullException(nameof(routeKey));
+            }
+
+            if (values == null)
+            {
+                throw new ArgumentNullException(nameof(values));
+            }
+
+            object obj;
+            if (values.TryGetValue(routeKey, out obj))
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Mvc.Core/Routing/KnownRouteValueEndpointMatchConstraint.cs]: out var obj

File: src/Microsoft.AspNetCore.Mvc.Core/Routing/KnownRouteValueEndpointMatchConstraint.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            {
+                var value = obj as string;
+                if (value != null)
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Mvc.Core/Routing/KnownRouteValueEndpointMatchConstraint.cs]: obj is string value

File: src/Microsoft.AspNetCore.Mvc.Core/Routing/KnownRouteValueEndpointMatchConstraint.cs [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            if (valuesCollection == null ||
+                version != valuesCollection.Version)
+            {
+                var values = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                for (var i = 0; i < actionDescriptors.Items.Count; i++)
+                {
+                    var action = actionDescriptors.Items[i];
+
+                    string value;
+                    if (action.RouteValues.TryGetValue(routeKey, out value) &&
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Mvc.Core/Routing/KnownRouteValueEndpointMatchConstraint.cs]: out var value

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs [hunk 7]
```diff
@@ -239,7 +248,7 @@ private bool MatchRouteValue(ActionDescriptor action, MvcEndpointInfo endpointIn
                     //
                     // REVIEW: This is really ugly
                     if (endpointInfo.Constraints.TryGetValue(routeKey, out var constraint)
-                        && !constraint.Match(new DefaultHttpContext() { RequestServices = _serviceProvider }, new DummyRouter(), routeKey, new RouteValueDictionary(action.RouteValues), RouteDirection.IncomingRequest))
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs]: Delete DummyRouter class (I think it is in this file)

**Comment by kichalla** [comment]
Closing this PR as it no longer valid after @rynowak RouteConstraintReference reaction changes.

**Pull request closed without merging**